### ISSUE TITLE
"Sign Up" creates user in Auth0 but not in Drupal

### ISSUE THREAD

**Issue by margyly** [auth0/auth0-drupal#110]
When I use the "Sign Up" option and enter an email that doesn't exist in the Auth0 database, the module creates the user in Auth0 but it doesn't create the user in the Drupal database. What is supposed to happen?

Desired result: Module creates a user account in the Drupal database, connected to the Auth0 user record.

**Comment by joshcanhelp** [comment]
Sorry for the trouble @margyly. Is there anything relevant in the Drupal logs? There is pretty thorough logging in the plugin.

That said, we're working on a new version of this module and will be going through some rigorous functional testing as part of that release. The `master` branch here has the alpha version that's slated for testing, might be worth trying that out.

**Comment by joshcanhelp** [comment]
@margyly - Actually, hold off on the master branch, there are a few fixes that need to be made there.

**Comment by joshcanhelp** [comment]
`master` branch is now in good shape.

I tested this and the user is created in both Auth0 and Drupal so I'm not sure what's happening on your end. If you can let me know what error message you're seeing and and relevant Drupal logs, that would be helpful. 

Thanks!

**Comment by margyly** [comment]
Well, damn, I've been using version 5.0.6 of the module! I changed the line in composer.json to:

"auth0/auth0-php": "5.1",

Now, when I look in Drupal, it shows: Version: 8.x-2.0.

**Comment by joshcanhelp** [comment]
@margyly - I'm going to "officially" release what's on the `master` branch here tomorrow as `2.1` here. I'll also sync that on Drupal to make a `8.x-2.1` release there. If you can wait until then, that's probably best. Otherwise, you'll need to manually check this repo out. 

In the Composer file, you don't want to adjust `auth0/auth0-php`, that's a library that the Drupal module depends on and it will update automatically. The `drupal/auth0` entry should be something like `^2.0` but that repo won't be up-to-date until tomorrow.

**Comment by margyly** [comment]
Sounds good. Thanks!

**Comment by margyly** [comment]
Can I reopen this, @joshcanhelp ? I'm now using version 8.x-2.1, but it still not creating new users in Drupal. From my site's home page at https://2061-auth0-signup-qbzpkxa-2gsgzx7spxgoy.us.platform.sh/, I choose to "Sign Up", enter an email address that does not exist in the Auth0 database, enter a password, and click "Sign Up". I return to the home page where I see the message "There was a problem logging you in." In Auth0 I see the new user account. In Drupal I do not. In the database log in Drupal I see the following log message:

```
Type: auth0_controller
Date: Sunday, September 30, 2018 - 8:32
User: Anonymous (not verified)
Location: https://2061-auth0-signup-qbzpkxa-2gsgzx7spxgoy.us.platform.sh/auth0/callback?error=unauthorized&error_description=Please%20verify%20your%20email%20before%20logging%20in.&state=5bb16b22e6d100.89201862
Referrer: https://2061-auth0-signup-qbzpkxa-2gsgzx7spxgoy.us.platform.sh/
Message: Failed to validate JWT: Wrong number of segments
Severity: Error
```
I've tried playing with the settings at /admin/config/auth0/advanced but no combination has worked. I'm currently trying:

* Allow user signup = checked
* Send a Refresh Token in the Signin Event for offline access = unchecked
* Redirect login for SSO = checked
* Requires verified email = unchecked
* Link Auth0 logins to Drupal users by email address = checked
* Auto Register Auth0 users (ignore site registration settings) = checked

Can you recommend settings?

I'd like to require that a new user verify their email address, but it's not clear how that will work. (What will they see on the Drupal site that tells them to check their email? Where would I configure that message?) 

Tl;dr: I can log in with an existing Auth0 account but I can't create a new one from Drupal.

**Comment by joshcanhelp** [comment]
@margyly - that's saying that the ID token received is not formatted properly, wouldn't be anything in settings that would do that. I tested everything straight through several times before releasing that so I don't think it's a problem with the module, also have not heard any other reports. 

It's failing here:

https://github.com/auth0/auth0-drupal/blob/master/src/Controller/AuthController.php#L458

... using this method specifically:

https://github.com/auth0/auth0-drupal/blob/master/src/Util/AuthHelper.php#L107

I just tried signing up with the Auth0 login form and verified email turned on and everything works as expected. When you sign up with the username/password, you get a message that your email needs to be verified, like so:

![screenshot 2018-10-02 09 05 34](https://user-images.githubusercontent.com/855223/46361443-86597f00-c622-11e8-88d3-1d78f53bc031.png)

You should already have a verify email link in your inbox but the Resend link works as well.

**Comment by margyly** [comment]
So what's failing is entirely within the module and there's nothing I could be doing to break it? Doesn't that sound like a problem with the module?

--
Margy Levine Young

**Comment by joshcanhelp** [comment]
What's failing could be a configuration on your server that's not playing well with the module. Without more information about why it's failing (e.g. why it's not seeing a valid ID token), I'm not sure what more I can do here. It very well could be an issue with the module but I have no way to replicate it, hence no way to fix it. If you can give me more information about the state of your application when that's happening (what is the ID token it's trying to check? is it empty? why?), I'd be happy to keep looking.

**Comment by margyly** [comment]
Wow, that's a very kind offer! Thanks! I know how annoying it is to try to fix something you can't replicate. 

Looking further, I see that the Drupal account doesn't get created until AFTER the user has clicked through on the confirmation email. Once I do that, it's created the account. I thought that it would create an unverified account in Drupal, and then verify it. 

SO, the error is mine! Woohoo! Next steps are getting the Drupal account to grab the username and other fields from Auth0 when creating the account, and having the user return to different pages depending on which application they logged in from.

Thanks again! 

P.S. If you are creating documentation for the module anywhere I'd be happy to contribute to it. I've written a bunch of "For Dummies" books and can explain the heck out of things.
